libnvme/sysfs.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Gris Ge <fge@redhat.com>
+ */
+
+#include <sys/types.h>
+#include <dirent.h>
+#include <errno.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <limits.h>
+
+#include <libnvme/libnvme.h>
+
+#include "sysfs.h"
+#include "utils.h"
+#include "ptr_array.h"
+
+#define _SYSFS_NVME_CLASS_PATH		"/sys/class/nvme"
+#define _SYSFS_BLK_PATH			"/sys/block"
+#define _UINT32_MAX_STR_LEN		11
+/* ^ The UINT32_MAX 4294967295 only takes 10 chars. */
+
+int _sysfs_nvme_ctrl_dev_paths_get(char *err_msg,
+				   const char ***nvme_ctrl_dev_paths,
+				   uint32_t *count)
+{
+	int rc = NVME_OK;
+	DIR *dir = NULL;
+	struct dirent *dp = NULL;
+	int errno_save = 0;
+	const char *nvme_ctrl_name = NULL;
+	char *nvme_ctrl_dev_path = NULL;
+	struct ptr_array *pa = NULL;
+	char *tmp_str = NULL;
+	uint32_t i = 0;
+
+	assert(nvme_ctrl_dev_paths != NULL);
+	assert(count != NULL);
+
+	*nvme_ctrl_dev_paths = NULL;
+	*count = 0;
+
+	pa = ptr_array_new(0);
+	if (pa == NULL) {
+		_nvme_err_msg_set(err_msg, "No memory");
+		rc = NVME_ERR_NO_MEMORY;
+		goto out;
+	}
+
+	dir = opendir(_SYSFS_NVME_CLASS_PATH);
+	if (dir == NULL) {
+		errno_save = errno;
+		if (errno_save != ENOENT)
+			_nvme_err_msg_set(err_msg, "Cannot open %s: error %d",
+					  _SYSFS_NVME_CLASS_PATH, errno_save);
```

Comment:
Not setting `rc`, thus `rc == NVME_OK`, so when we get to the `out` we think we were successful.